core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -497,30 +512,51 @@ class Partition(val topicPartition: TopicPartition,
    * @return true if the high watermark has been updated
    */
   def maybeExpandIsr(replicaId: Int, logReadResult: LogReadResult): Boolean = {
-    inWriteLock(leaderIsrUpdateLock) {
+    var hwIncremented = false
+    val (leaderReplica, needsIsrUpdate) = inReadLock(leaderIsrUpdateLock) {
```

Comment:
Could we use `StampedLock.asReadLock` or something like that for better performance?

Reply:
@ijuma Thanks for the review, Since StampedLocks are not reentrant, the code may become harder to maintain. So I left the lock as ReentrantReadWriteLock, even though as you say it doesn't perform as well for this case. Since updates are relatively infrequent, perhaps that is ok?